File: android/pigeon/src/main/kotlin/com/marcelocantos/pigeon/crypto/CredentialStore.kt
// Copyright 2026 Marcelo Cantos
// SPDX-License-Identifier: Apache-2.0

package com.marcelocantos.pigeon.crypto

import java.io.File
import java.nio.file.Files
import java.nio.file.attribute.PosixFilePermission
import java.time.Instant

/** Thrown when no credential is currently stored. */
class NoCredentialException : Exception("pigeon: no credential stored")

/**
 * A uniform interface for persisting a [PairingArtifact] across
 * processes/restarts. Backing implementations are platform-specific
 * (EncryptedSharedPreferences on Android, file on JVM/desktop); the
 * interface is the same.
 */
interface CredentialStore {
    /** Replaces any existing artifact. */
    fun save(artifact: PairingArtifact)
    /** Returns the stored artifact or throws [NoCredentialException]. */
    fun load(): PairingArtifact
    /** Removes the stored artifact. No-op if absent. */
    fun delete()
    /**
     * Convenience: loads and reports whether the stored artifact's
     * `expiresAt` is in the past. Throws [NoCredentialException] when
     * nothing is stored.
     */
    fun isExpired(): Boolean
}

/**
 * File-backed credential store. Stores the artifact's canonical JSON
 * at the given path. On POSIX systems, restricts permissions to 0600;
 * on Windows the default ACLs apply.
 *
 * Suitable for JVM/desktop scenarios. Production deployments on
 * Android should use [EncryptedSharedPreferencesCredentialStore]
 * (which lives in the Android-specific module — not in this JVM
 * library because it requires `androidx.security`).
 */
class FileCredentialStore(
    val path: File,
    var now: () -> Instant = { Instant.now() },
) : CredentialStore {

    override fun save(artifact: PairingArtifact) {
        val parent = path.parentFile
        if (parent != null && !parent.exists()) {
            parent.mkdirs()
        }
        val tmp = File(path.absolutePath + ".tmp")
        tmp.writeText(artifact.toJson(), Charsets.UTF_8)
        if (!tmp.renameTo(path)) {
            tmp.delete()
            throw java.io.IOException("rename ${tmp.absolutePath} -> ${path.absolutePath}")
        }
        applyOwnerOnlyPermissions(path)
    }

    override fun load(): PairingArtifact {
        if (!path.exists()) throw NoCredentialException()
        val text = path.readText(Charsets.UTF_8)
        return PairingArtifact.fromJson(text)
    }

    override fun delete() {
        if (path.exists() && !path.delete()) {
            throw java.io.IOException("delete ${path.absolutePath}")
        }
    }

    override fun isExpired(): Boolean = load().isExpired(now())
}

private fun applyOwnerOnlyPermissions(file: File) {
    val supported = file.toPath().fileSystem.supportedFileAttributeViews().contains("posix")
    if (!supported) return
    val perms = setOf(PosixFilePermission.OWNER_READ, PosixFilePermission.OWNER_WRITE)
    Files.setPosixFilePermissions(file.toPath(), perms)
}

File: credential_store.go
// Copyright 2026 Marcelo Cantos
// SPDX-License-Identifier: Apache-2.0

package pigeon

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"time"
)

// ErrNoCredential is returned by CredentialStore.Load when no artifact
// has been saved.
var ErrNoCredential = errors.New("pigeon: no credential stored")

// CredentialStore is a uniform interface for persisting a
// PairingArtifact across processes/restarts. Backing implementations
// are platform-specific (file+OS-keyring on desktop, Keychain on iOS,
// EncryptedSharedPreferences on Android); the interface is the same.
//
// Save replaces any existing artifact. Load returns ErrNoCredential
// when nothing is stored. Delete removes the stored artifact (no-op
// if absent). IsExpired is a convenience that loads and checks
// expiry; it returns (false, ErrNoCredential) if nothing is stored.
type CredentialStore interface {
	Save(artifact *PairingArtifact) error
	Load() (*PairingArtifact, error)
	Delete() error
	IsExpired() (bool, error)
}

// FileCredentialStore is a reference implementation that persists the
// artifact as canonical JSON in a single file with 0600 permissions.
// Suitable for desktop scenarios. Production deployments on iOS or
// Android should use the Keychain or EncryptedSharedPreferences
// implementations in the Swift / Kotlin SDKs respectively.
type FileCredentialStore struct {
	// Path is the absolute file path where the artifact is stored.
	Path string
	// Now is injectable for testing; defaults to time.Now.
	Now func() time.Time
}

// NewFileCredentialStore returns a store backed by the given path. The
// parent directory is created on first Save.
func NewFileCredentialStore(path string) *FileCredentialStore {
	return &FileCredentialStore{Path: path}
}

func (s *FileCredentialStore) now() time.Time {
	if s.Now != nil {
		return s.Now()
	}
	return time.Now()
}

// Save writes the artifact to disk, replacing any existing file.
func (s *FileCredentialStore) Save(a *PairingArtifact) error {
	if a == nil {
		return errors.New("pigeon: cannot save nil artifact")
	}
	data, err := a.Marshal()
	if err != nil {
		return fmt.Errorf("marshal artifact: %w", err)
	}
	if err := os.MkdirAll(filepath.Dir(s.Path), 0o700); err != nil {
		return fmt.Errorf("create credential dir: %w", err)
	}
	tmp := s.Path + ".tmp"
	if err := os.WriteFile(tmp, data, 0o600); err != nil {
		return fmt.Errorf("write credential: %w", err)
	}
	if err := os.Rename(tmp, s.Path); err != nil {
		_ = os.Remove(tmp)
		return fmt.Errorf("rename credential: %w", err)
	}
	return nil
}

// Load reads the artifact from disk. Returns ErrNoCredential if the
// file does not exist.
func (s *FileCredentialStore) Load() (*PairingArtifact, error) {
	data, err := os.ReadFile(s.Path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, ErrNoCredential
		}
		return nil, fmt.Errorf("read credential: %w", err)
	}
	return UnmarshalPairingArtifact(data)
}

// Delete removes the stored artifact. No-op if no artifact is stored.
func (s *FileCredentialStore) Delete() error {
	if err := os.Remove(s.Path); err != nil && !errors.Is(err, os.ErrNotExist) {
		return fmt.Errorf("delete credential: %w", err)
	}
	return nil
}

// IsExpired loads the artifact and reports whether its ExpiresAt is in
// the past at the store's current time.
func (s *FileCredentialStore) IsExpired() (bool, error) {
	a, err := s.Load()
	if err != nil {
		return false, err
	}
	return a.IsExpired(s.now()), nil
}
